File: frontend/api/calculations/excel_export.py
"""Excel export for ФЭМ-ассистент"""

from openpyxl import Workbook
from openpyxl.styles import Font, Alignment, PatternFill, Border, Side
from io import BytesIO

from .models import Assumptions
from .engine import run_calculation

HEADER_FILL = PatternFill(start_color="4472C4", end_color="4472C4", fill_type="solid")
HEADER_FONT = Font(bold=True, color="FFFFFF", size=10)
LABEL_FONT = Font(bold=True, size=10)
NORMAL_FONT = Font(size=10)
THIN_BORDER = Border(
    left=Side(style="thin"),
    right=Side(style="thin"),
    top=Side(style="thin"),
    bottom=Side(style="thin"),
)


def _style_header(ws, row, cols):
    for c in range(1, cols + 1):
        cell = ws.cell(row=row, column=c)
        cell.fill = HEADER_FILL
        cell.font = HEADER_FONT
        cell.alignment = Alignment(horizontal="center", wrap_text=True)
        cell.border = THIN_BORDER


def _style_row(ws, row, cols, bold=False):
    for c in range(1, cols + 1):
        cell = ws.cell(row=row, column=c)
        cell.border = THIN_BORDER
        cell.font = LABEL_FONT if bold else NORMAL_FONT
        if c == 1:
            cell.alignment = Alignment(horizontal="left")
        else:
            cell.alignment = Alignment(horizontal="center")


def generate_excel(a: Assumptions, data: dict) -> bytes:
    wb = Workbook()

    # Sheet 1: Assumptions
    ws = wb.active
    ws.title = "Assumptions"
    ws.cell(1, 1, "Параметр").font = HEADER_FONT
    ws.cell(1, 2, "Значение").font = HEADER_FONT
    rows_a = [
        ("Целевая выручка в день", a.target_revenue_per_day),
        ("Средний чек", a.avg_check),
        ("Заказов в день", a.orders_per_day),
        ("Ramp-up мес 1", a.ramp_month_1),
        ("Ramp-up мес 2", a.ramp_month_2),
        ("Ramp-up мес 3+", a.ramp_month_3),
        ("Себестоимость заказа", a.cost_per_order),
        ("ФОТ", a.payroll),
        ("Эквайринг %", a.acquiring_pct),
        ("Аренда", a.rent),
        ("Маркетинг", a.marketing),
        ("Логистика на заказ", a.logistics_per_order),
        ("Прочие расходы", a.other_expenses),
        ("Регистрация ИП", a.reg_ip),
        ("Оборудование", a.equipment),
        ("Сайт", a.website),
        ("Резервный фонд", a.reserve_fund),
        ("Налоговая ставка", a.tax_rate),
        ("Дни запаса", a.inventory_days),
        ("Отсрочка поставщика", a.supplier_deferral),
        ("Отсрочка клиента", a.customer_deferral),
    ]
    for i, (label, val) in enumerate(rows_a, 2):
        ws.cell(i, 1, label).font = NORMAL_FONT
        ws.cell(i, 2, val).font = NORMAL_FONT
    ws.column_dimensions["A"].width = 25
    ws.column_dimensions["B"].width = 15

    months = [f"Мес {i}" for i in range(1, 37)]

    def write_table(ws, title, rows_data, row_start=1):
        ws.cell(row_start, 1, title).font = Font(bold=True, size=12)
        for c_idx, label in enumerate(["Показатель"] + months, 1):
            cell = ws.cell(row_start + 1, c_idx, label)
            cell.font = HEADER_FONT
            cell.fill = HEADER_FILL
            cell.alignment = Alignment(horizontal="center", wrap_text=True)
            cell.border = THIN_BORDER
        for r_idx, (label, values) in enumerate(rows_data, row_start + 2):
            ws.cell(r_idx, 1, label).font = LABEL_FONT
            ws.cell(r_idx, 1).border = THIN_BORDER
            for c_idx, val in enumerate(values, 2):
                cell = ws.cell(r_idx, c_idx, round(val, 2))
                cell.font = NORMAL_FONT
                cell.border = THIN_BORDER
                cell.alignment = Alignment(horizontal="center")
        return row_start + 2 + len(rows_data) + 1

    # Sheet 2: P&L
    ws2 = wb.create_sheet("P&L")
    pnl = data["pnl"]
    pnl_data = [
        ("Выручка", pnl["revenue"]),
        ("Эквайринг", pnl["acquiring"]),
        ("Чистая выручка", pnl["net_revenue"]),
        ("Себестоимость", pnl["cost_of_goods"]),
        ("ФОТ", pnl["payroll"]),
        ("Логистика", pnl["logistics"]),
        ("Маркетинг", pnl["marketing"]),
        ("Аренда", pnl["rent"]),
        ("Прочие", pnl["other"]),
        ("Итого расходов", pnl["total_expenses"]),
        ("Прибыль до налога", pnl["profit_before_tax"]),
        ("Налог", pnl["tax"]),
        ("Чистая прибыль", pnl["net_profit"]),
        ("Накопленная прибыль", pnl["cumulative_profit"]),
    ]
    write_table(ws2, "Отчёт о прибылях и убытках", pnl_data)

    # Sheet 3: Cash Flow
    ws3 = wb.create_sheet("Cash Flow")
    cf = data["cashflow"]
    cf_data = [
        ("Поступления", cf["operating_inflow"]),
        ("Платежи (-)", cf["operating_outflow"]),
        ("Чистый опер. поток", cf["net_operating"]),
        ("Инвестиции", cf["investing"]),
        ("Финансирование", cf["financing"]),
        ("ДС на начало", cf["cash_start"]),
        ("ДС на конец", cf["cash_end"]),
        ("ЧДПС", cf["net_cashflow"]),
    ]
    write_table(ws3, "Бюджет движения денежных средств", cf_data)

    # Sheet 4: Balance Sheet
    ws4 = wb.create_sheet("Balance Sheet")
    bs = data["balance_sheet"]
    bs_data = [
        ("Денежные средства", bs["cash"]),
        ("Запасы", bs["inventory"]),
        ("Основные средства", bs["fixed_assets"]),
        ("Итого активы", bs["total_assets"]),
        ("Капитал", bs["capital"]),
        ("Кредиторская задолж.", bs["accounts_payable"]),
        ("Итого пассивы", bs["total_liabilities"]),
    ]
    write_table(ws4, "Баланс", bs_data)

    # Sheet 5: Ratios
    ws5 = wb.create_sheet("Ratios")
    r = data["ratios"]
    ws5.cell(1, 1, "Показатель").font = HEADER_FONT
    ws5.cell(1, 2, "Значение").font = HEADER_FONT
    ws5.cell(1, 3, "Норма").font = HEADER_FONT
    _style_header(ws5, 1, 3)
    ratio_rows = [
        ("ROE (рентабельность капитала)", r["ROE"], "> 15%"),
        ("ROCE (рентабельность капитала)", r["ROCE"], "> 12%"),
        ("OPM (операционная маржа)", r["OPM"], "> 10%"),
        ("GPM (валовая маржа)", r["GPM"], "> 20%"),
        ("Проверка баланса", r["BalanceCheck"], "0"),
    ]
    for i, (label, val, norm) in enumerate(ratio_rows, 2):
        ws5.cell(i, 1, label).font = NORMAL_FONT
        ws5.cell(i, 2, val).font = NORMAL_FONT
        ws5.cell(i, 3, norm).font = NORMAL_FONT
    ws5.column_dimensions["A"].width = 35
    ws5.column_dimensions["B"].width = 15
    ws5.column_dimensions["C"].width = 10

    output = BytesIO()
    wb.save(output)
    output.seek(0)
    return output.getvalue()

File: frontend/src/lib/excel-export.ts
/** Excel export for ФЭМ-ассистент */
import ExcelJS from 'exceljs';
import { runCalculation } from './engine';

const HEADER_FILL = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF4472C4' } } as const;
const HEADER_FONT = { bold: true, color: { argb: 'FFFFFFFF' }, size: 10 };

export async function generateExcel(a: Record<string, any>): Promise<Buffer> {
  const data = runCalculation(a);
  const wb = new ExcelJS.Workbook();

  // Sheet 1: Assumptions
  const ws = wb.addWorksheet('Assumptions');
  ws.columns = [
    { header: 'Параметр', key: 'param', width: 30 },
    { header: 'Значение', key: 'value', width: 20 },
  ];
  const rowsA = [
    ['Месяцев', a.months || 36],
    ['Инфляция месячная', a.monthly_inflation || 0],
    ['Товаров', (a.products || []).length],
    ['Налоговая система', a.taxSystem || a.tax_system || 'не указана'],
    ['Тип УСН', a.usnType || a.usn_type || 'income_minus_expenses'],
    ['ФОТ', parseFloat(a.payroll || '0')],
    ['Аренда', parseFloat(a.rent || '0')],
    ['Прочие расходы', parseFloat(a.other_expenses || '0')],
    ['Фикс. взносы ИП', parseFloat(a.ip_fixed_contrib || a.insurance_fixed || '59700')],
    ['Доп. взносы ИП 1%', parseFloat(a.ip_extra_contrib || a.ipExtraContrib || '0')],
    ['Эквайринг %', parseFloat(a.acquiring_pct || a.acquiringPct || '1.5')],
    ['Дни запаса', parseFloat(a.inventory_days || a.inventoryDays || '30')],
    ['Отсрочка поставщика (дн)', parseFloat(a.supplier_deferral || a.supplierDeferral || '0')],
    ['Отсрочка клиента (дн)', parseFloat(a.customer_deferral || a.customerDeferral || '0')],
  ];
  for (const [param, val] of rowsA) {
    ws.addRow([param, val]);
  }

  // Style header
  const r1 = ws.getRow(1);
  r1.font = HEADER_FONT as any;
  r1.fill = HEADER_FILL as any;
  r1.alignment = { horizontal: 'center' };

  const months = Array.from({ length: Math.min(36, data.pnl.revenue.length) }, (_, i) => `Мес ${i + 1}`);

  function writeTable(sheet: ExcelJS.Worksheet, title: string, rowsData: [string, number[]][]): void {
    sheet.addRow([]);
    sheet.addRow([title]);
    const titleRow = sheet.lastRow!;
    titleRow.font = { bold: true, size: 12 };

    const headerRow = sheet.addRow(['Показатель', ...months]);
    headerRow.font = HEADER_FONT as any;
    headerRow.fill = HEADER_FILL as any;
    headerRow.alignment = { horizontal: 'center' };

    for (const [label, values] of rowsData) {
      const row = sheet.addRow([label, ...values.slice(0, months.length).map(v => Math.round(v * 100) / 100)]);
      row.getCell(1).font = { bold: true };
      for (let c = 2; c <= months.length + 1; c++) {
        row.getCell(c).alignment = { horizontal: 'center' };
      }
    }
  }

  // Sheet 2: P&L
  const ws2 = wb.addWorksheet('P&L');
  const pnlData: [string, number[]][] = [
    ['Выручка', data.pnl.revenue],
    ['Скидки', data.pnl.discounts_total || []],
    ['Чистая выручка', data.pnl.net_revenue],
    ['Закуп', data.pnl.purchase_cogs],
    ['Переменные расходы', data.pnl.variable_extra],
    ['Валовая прибыль', data.pnl.gross_profit],
    ['Валовая маржа %', data.pnl.gross_margin],
    ['Постоянные расходы', data.pnl.fixed_costs],
    ['ФОТ + налоги', data.pnl.payroll_total],
    ['Итого постоянные', data.pnl.total_fixed],
    ['EBITDA', data.pnl.ebitda],
    ['Маржа EBITDA %', data.pnl.ebitda_margin],
    ['Амортизация', data.pnl.amortization],
    ['Налог', data.pnl.tax],
    ['Чистая прибыль', data.pnl.net_profit],
    ['Накопленная прибыль', data.pnl.cumulative_profit],
  ];
  writeTable(ws2, 'Отчёт о прибылях и убытках (P&L)', pnlData);

  // Sheet 3: Cash Flow
  const ws3 = wb.addWorksheet('Cash Flow');
  const cfData: [string, number[]][] = [
    ['Поступления', data.cashflow.operating_inflow],
    ['Платежи (-)', data.cashflow.operating_outflow],
    ['Чистый опер. поток', data.cashflow.net_operating],
    ['Инвестиции', data.cashflow.investing],
    ['Финансирование', data.cashflow.financing],
    ['ДС на начало', data.cashflow.cash_start],
    ['ДС на конец', data.cashflow.cash_end],
    ['ЧДПС', data.cashflow.net_cashflow],
  ];
  writeTable(ws3, 'Бюджет движения денежных средств', cfData);

  // Sheet 4: Balance Sheet
  const ws4 = wb.addWorksheet('Balance Sheet');
  const bsData: [string, number[]][] = [
    ['Денежные средства', data.balance_sheet.cash],
    ['Запасы', data.balance_sheet.inventory],
    ['Дебиторка', data.balance_sheet.receivables],
    ['Основные средства', data.balance_sheet.fixed_assets],
    ['Итого активы', data.balance_sheet.total_assets],
    ['Капитал', data.balance_sheet.capital],
    ['Кредиторка', data.balance_sheet.accounts_payable],
    ['Итого пассивы', data.balance_sheet.total_liabilities],
  ];
  writeTable(ws4, 'Баланс', bsData);

  // Sheet 5: Ratios
  const ws5 = wb.addWorksheet('Ratios');
  ws5.columns = [
    { header: 'Показатель', key: 'ratio', width: 40 },
    { header: 'Значение', key: 'value', width: 15 },
    { header: 'Норма', key: 'norm', width: 10 },
  ];
  ws5.addRow(['ROE (рентабельность капитала)', data.ratios.ROE, '> 15%']);
  ws5.addRow(['OPM (операционная маржа)', data.ratios.OPM, '> 10%']);
  ws5.addRow(['GPM (валовая маржа)', data.ratios.GPM, '> 20%']);
  ws5.addRow(['Проверка баланса', data.ratios.BalanceCheck, '0']);

  if (data.break_even) {
    ws5.addRow(['Точка безубыточности (заказов/мес)', data.break_even.break_even_orders, '']);
    ws5.addRow(['Постоянные расходы/мес', data.break_even.monthly_fixed, '']);
    ws5.addRow(['Маржа на единицу', data.break_even.margin_per_unit, '']);
  }

  const r5 = ws5.getRow(1);
  r5.font = HEADER_FONT as any;
  r5.fill = HEADER_FILL as any;

  const buf = await wb.xlsx.writeBuffer();
  return Buffer.from(buf);
}
